### DISCUSSION TITLE
NEXT_PUBLIC environment variable not being read from shell

### DISCUSSION THREAD

**Discussion by barankyle** [vercel/next.js#13427] [Help]
I just upgraded to 9.4 to try to solve my client-side environment variable troubles, but that doesn't seem to have fixed it, either.

My deployment is on Kubernetes via Helm. I'm setting an ENV_VAR 'NEXT_PUBLIC_API_SERVER' as one of the values passed into Helm. It's definitely being set properly, because if I go into the pod and run `echo $NEXT_PUBLIC_API_SERVER`, I get the value I expect.

I'm using it in the client code in one place, which looks like this:

`export const apiUrl = process.env.NODE_ENV === 'production' ? process.env.NEXT_PUBLIC_API_SERVER : 'http://localhost:3030'`

This works fine when I'm testing locally, albeit when i swap 'production' for 'development'. But in production, process.env is empty in the code sent to the browser.

I don't have a .env file anywhere in this codebase, and dotenv is not in the package.json and is not installed.

What am I missing?

**Comment by malimccalla** [comment]
Hey @barankyle 

I believe I have a similar set up to you, my app sets `NEXT_PUBLIC_API_ENDPOINT` and is also deployed to a Kubernetes cluster.

Are you currently setting `NEXT_PUBLIC_API_SERVER` when you build your app with `next build`? I don't believe you need to pass it to your pod as it gets set when you build your next app. I think your problem may be that you are not setting the value at build time.

As you are using kubernetes I assume you have a docker image that builds a container to run your the next.js app. Do you mind sharing your Dockerfile? If I were you I would try building and running your container locally to see if the variables have been set correctly

**Reply by malimccalla** [reply]
If it helps, mine is working with this Dockerfile

```dockerfile
# Build phase
FROM node:latest as builder
WORKDIR /app

COPY ./package.json .
COPY ./yarn.lock .
COPY ./tsconfig.json .

# Will use cache unless any of the above files have changed
RUN yarn install

# Copy the source code to be built
COPY . .

ARG NEXT_PUBLIC_API_ENDPOINT
ENV NEXT_PUBLIC_API_ENDPOINT=$NEXT_PUBLIC_API_ENDPOINT

# executes `next build`
RUN yarn run build 

# Run phase
FROM node:lts-alpine
WORKDIR /app

COPY --from=builder /app/package.json .
COPY --from=builder /app/yarn.lock .

RUN yarn install --production

# Get the build output
COPY --from=builder /app/dist /app/dist
COPY --from=builder /app/.next /app/.next

# Copy over the public folder
COPY --from=builder /app/public /app/public

ENV NODE_ENV=production

EXPOSE 3000

CMD ["node", "dist/server.js"]
```

Take note of these two lines in particular.

```dockerfile
ARG NEXT_PUBLIC_API_ENDPOINT
ENV NEXT_PUBLIC_API_ENDPOINT=$NEXT_PUBLIC_API_ENDPOINT
```

When I build the docker image I pass in a docker `--build-arg` like this

```sh
docker build -t <docker_id>/<image_name> \
  --build-arg=NEXT_PUBLIC_API_ENDPOINT=https://api.example.com \
  -f ./path/to/Dockerfile .
```

which allows my image to pick up the `NEXT_PUBLIC_API_ENDPOINT` and then set it as an environment variable. This environment variable is then set when I build my next app with `next build`.

I use this Dockerfile for both staging and production so the value needs to be different for each. However if you only use your Dockerfile for production you can just hardcode the `ENV` in the dockerfile without having to worry about build args

**Reply by barankyle** [reply]
That might be the problem. We do very little in the Dockerfile:

```dockerfile
# this file generates a very large image, because it uses full node image on debian
# and because package.json is not optimized and pulls AFrame and three.js from github 
# this is ok in development and internal builds initially and once optimized we can switch to 2-stage builds

# not slim because we need github depedencies
FROM node:12.16

# Create app directory
WORKDIR /app

# to make use of caching, copy only package files and install dependencies
COPY package*.json /app/
#RUN  npm ci --verbose  # we should make lockfile or shrinkwrap then use npm ci for predicatble builds
RUN  npm install --no-progress --verbose

# Build Args, NOTE: supplied at build time, not runtime
#ARG API_SERVER_URL=http://localhost:3030

# copy then compile the code
COPY . .
RUN npm run build

EXPOSE 80
CMD [ "npm", "start" ]
```
Someone else on my team wrote this, but I noticed that they commented out setting an ARG (naming differences chalked up to things having changed). Maybe we need to do that and set it as an ENV.

**Reply by malimccalla** [reply]
Yes I would try adding in:

```dockerfile
ENV NEXT_PUBLIC_API_SERVER_URL=http://localhost:3030
```

to your dockerfile. Your variable will then be accessible in your app with `process.env.NEXT_PUBLIC_API_SERVER_URL`

**Reply by exocode** [reply]
> ```dockerfile
> ARG NEXT_PUBLIC_API_ENDPOINT
> ENV NEXT_PUBLIC_API_ENDPOINT=$NEXT_PUBLIC_API_ENDPOINT
> ```
>
> I use this Dockerfile for both staging and production so the value needs to be different for each. However if you only use your Dockerfile for production you can just hardcode the `ENV` in the dockerfile without having to worry about build args

Yes, that's weired... we "bake" VARIABLES into the image.. 
So we need different images for each stage? Does this not destroy the purpose of images (having exact same setup for all stages)?
How do you handle such a case when you have different stages? Production, QA, Testing,...

**Comment by frankrod** [comment]
Hello, I have sort of the same issue but I'm not using kubernetes or helm. This is my case:

nextjs v9.4.4 with a custom server where I use dotenv as a devdepency, I have a .env file locally(it's in .gitignore).
And I would like to generate my production build by just typing:
`NEXT_PUBLIC_MY_VAR=some-text next build`
If I have dotenv installed I cannot set it that way it is always undefined but if I uninstalled dotenv it works.
Any way you know how can I set my NEXT_PUBLIC_  from shell without uninstalling dotenv ?

**Comment by rene-demonsters** [comment]
The problem that @barankyle is experiencing is caused by Automatic Runtime Optimization as described here: https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration. The environment variables page (unfortunately) makes no mention of this: https://nextjs.org/docs/basic-features/environment-variables

We just spend a few hours debugging this ourselves. It might be good to add this mention to the environment variables page as well.

So the difference lies between build-time environment variables and runtime environment variables.  The reason that  @malimccalla's configuration works, is indeed because he has ARGS/ENV in his docker file, thus passing build-time environment variables. Those are then baked in by the optimization, they're no longer read at runtime.

**Comment by Vadorequest** [comment]
Could be also related to the fact that you need to add non-public variables you want to use at runtime that must be added to the build configuration.

See https://github.com/UnlyEd/next-right-now/blob/v2-mst-aptd-at-lcz-sty/next.config.js#L19-L34

**Comment by T04435** [comment]
Hope this helps:

https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser

Also make sure you restart your server

**Comment by olamide226** [comment]
The key gotcha here is that `NEXT_PUBLIC_*` values are not read by the browser from the pod environment at request time. They are substituted into the client bundle during `next build`. So Kubernetes/Helm can show the right value inside the running pod:

```bash
kubectl exec deploy/web -- printenv NEXT_PUBLIC_API_SERVER
```

…but the already-built browser JS will still contain whatever value existed when the image was built.

If the goal is a single image promoted through dev/staging/prod, build args solve a different problem. They let you bake a value into the image at build time. They do not give you runtime browser config.

This Kubernetes case is one of the reasons I’ve been working on REP:

https://github.com/RuachTech/rep

REP’s pattern is:

- stop using `NEXT_PUBLIC_*` for values that must change after `next build`
- pass `REP_PUBLIC_*` vars at container runtime instead
- run a small gateway in front of Next.js
- let the gateway inject an inert JSON payload into HTML
- read those values from client code synchronously

For example, instead of:

```ts
export const apiUrl = process.env.NEXT_PUBLIC_API_SERVER
```

use:

```ts
'use client'

import { rep } from '@rep-protocol/sdk'

export const apiUrl = rep.get('API_SERVER')
```

Then set the value through Kubernetes/Helm as a runtime env var:

```yaml
env:
  - name: REP_PUBLIC_API_SERVER
    value: "https://api.staging.example.com"
  - name: REP_PUBLIC_ENV_NAME
    value: "staging"
```

A minimal standalone Next.js container can run like this:

```dockerfile
# next.config.js should use: output: 'standalone'

FROM node:22-alpine AS builder
WORKDIR /app
COPY package.json package-lock.json* ./
RUN npm ci
COPY . .
RUN npm run build

FROM node:22-alpine AS runner
WORKDIR /app

COPY --from=ghcr.io/ruachtech/rep/gateway:latest /usr/local/bin/rep-gateway /usr/local/bin/rep-gateway

COPY --from=builder /app/.next/standalone ./
COPY --from=builder /app/.next/static ./.next/static
COPY --from=builder /app/public ./public

COPY docker-entrypoint.sh /docker-entrypoint.sh
RUN chmod +x /docker-entrypoint.sh

EXPOSE 8080
ENTRYPOINT ["/docker-entrypoint.sh"]
```

```sh
#!/bin/sh
set -e

NODE_ENV=production node /app/server.js &

exec /usr/local/bin/rep-gateway \
  --mode proxy \
  --port "${REP_GATEWAY_PORT:-8080}" \
  --upstream "localhost:3000"
```

Build once:

```bash
docker build -t my-next-app:runtime-config .
```

Run locally like a staging pod:

```bash
docker run --rm -p 8080:8080 \
  -e REP_PUBLIC_API_SERVER=https://api.staging.example.com \
  -e REP_PUBLIC_ENV_NAME=staging \
  my-next-app:runtime-config
```

Run the same image like production:

```bash
docker run --rm -p 8080:8080 \
  -e REP_PUBLIC_API_SERVER=https://api.example.com \
  -e REP_PUBLIC_ENV_NAME=production \
  my-next-app:runtime-config
```

And in Helm/Kubernetes, the useful shape is:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: web
spec:
  template:
    spec:
      containers:
        - name: web
          image: my-next-app:runtime-config
          ports:
            - containerPort: 8080
          env:
            - name: REP_PUBLIC_API_SERVER
              valueFrom:
                configMapKeyRef:
                  name: web-runtime-config
                  key: apiServer
            - name: REP_PUBLIC_ENV_NAME
              value: production
```

The important boundary is that `REP_PUBLIC_*` is still public browser config. Use it for API base URLs, OAuth client IDs, feature flags, Sentry DSNs, etc. Keep true secrets as normal server-only env vars and read them only from server-side code.
